kube-core/src/cel.rs

```diff
@@ -280,3 +282,137 @@ pub fn merge_properties(s: &mut Schema, merge: &mut Schema) {
         }
     }
 }
+
+/// ListType represents x-kubernetes merge strategy for list.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum ListMerge {
+    /// Atomic represents a list, where entire list is replaced during merge. At any point in time, a single manager owns the list.
+    Atomic,
+    /// Set applies to lists that include only scalar elements. These elements must be unique.
+    Set,
+    /// Map applies to lists of nested types only. The key values must be unique in the list.
+    Map(Vec<String>),
+}
+
+/// MapMerge represents x-kubernetes merge strategy for map.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum MapMerge {
+    /// Atomic represents a map, which can only be entirely replaced by a single manager.
+    Atomic,
+    /// Granular represents a map, which supports separate managers updating individual fields.
+    Granular,
+}
+
+/// StructMerge represents x-kubernetes merge strategy for struct.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum StructMerge {
+    /// Atomic represents a struct, which can only be entirely replaced by a single manager.
+    Atomic,
+    /// Granular represents a struct, which supports separate managers updating individual fields.
+    Granular,
+}
```

Comment:
this could potentially be a common `untagged` enum that we put inside the MergeStrategy and MergeStrategy (where they are reused). not sure if this is smart though. wdyt?

i was trying to find information on upstream stuff and ended up finding even more variants (atomic/associative/separable) in https://github.com/kubernetes-sigs/structured-merge-diff/blob/e8775feb43c2ee5a4883d791bce62d2669ac13a6/schema/elements.go#L88-L101 . possibly that is a more low level primitive on the kubernetes side.

Reply:
I mainly used official documentation for the functionality. Looking into https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/555-server-side-apply/README.md, there seems to be no mention of the associative type, so I’m not sure how and if it works.

But the enum leaves options to add values in the future, if needed?

Reply:
yeah, your setup is sensible, my comment is just a stray thought tbh. resolving.